src/cockpit/polkit.py

```diff
@@ -0,0 +1,139 @@
+import asyncio
+import locale
+import logging
+import os
+import pwd
+
+from cockpit._vendor.systemd_ctypes import bus
+
+# that path is valid on at least Debian, Fedora/RHEL, and Arch
+HELPER_PATH = '/usr/lib/polkit-1/polkit-agent-helper-1'
+
+logger = logging.getLogger(__name__)
+
+
+# https://www.freedesktop.org/software/polkit/docs/latest/eggdbus-interface-org.freedesktop.PolicyKit1.AuthenticationAgent.html
+class org_freedesktop_PolicyKit1_AuthenticationAgent(bus.Object):
+    def __init__(self, responder):
+        super().__init__()
+        self.responder = responder
+
+    @bus.Interface.Method('', ['s', 's', 's', 'a{ss}', 's', 'a(sa{sv})'])
+    async def begin_authentication(self, action_id, message, icon_name, details, cookie, identities):
```

Comment:
> The authentication agent should not return until after authentication is complete. If the user dismisses the authentication dialog, the authentication agent should return the [org.freedesktop.PolicyKit1.Error.Cancelled](https://www.freedesktop.org/software/polkit/docs/latest/eggdbus-interface-org.freedesktop.PolicyKit1.Authority.html#eggdbus-constant-Error.org.freedesktop.PolicyKit1.Error.Cancelled) error.

Interesting.  This deserves a comment, since the name ("begin") suggests that this function ought not to be async.

Reply:
Right, added a comment.